api/v2/api.go

```diff
@@ -638,13 +638,28 @@ func sortSilences(sils open_api_models.GettableSilences) {
 	})
 }
 
+// gettableSilenceMatchesFilterLabels returns true if
+// a given silence matches a list of matchers.
+// A silence matches a filter (list of matchers) if
+// for all matchers in the filter, there exists a matcher in the silence
+// such that their names, types, and values are equivalent.
 func gettableSilenceMatchesFilterLabels(s open_api_models.GettableSilence, matchers []*labels.Matcher) bool {
-	sms := make(map[string]string)
-	for _, m := range s.Matchers {
-		sms[*m.Name] = *m.Value
+	for _, matcher := range matchers {
+		found := false
+		for _, m := range s.Matchers {
+			if matcher.Name == *m.Name &&
+				(matcher.Type == labels.MatchEqual && !*m.IsRegex || matcher.Type == labels.MatchRegexp && *m.IsRegex) &&
+				matcher.Value == *m.Value {
+				found = true
+				break
+			}
+		}
+		if !found {
+			return false
+		}
 	}
 
-	return matchFilterLabels(matchers, sms)
```

Comment:
If the new logic you have here were refactored into `matchFilterLabels`, would it also improve alert filtering (since `matchFilterLabels` seems to be used by that endpoint too)?

Reply:
Hmm, I think it wouldn't. IMO we should apply different logic to filter alerts or silences.

For instance, suppose we'd like to filter out alerts coming from `*.example.com` -- we can use regexp `instance=~".*.example.com"` for this purpose. In this case, we should rather apply the existing `matchFilterLabels` logic, which checks whether an alert label matches with the given matcher by using regexp.

On the other hand, unlike alerts, a label of silence may contain regex like `instance=~".*.example.com"` (silence all alerts from *.example.com). In this case, we should _not_ use regexp to check whether the silence label matches with the given matcher, but apply a simple string comparison.

Reply:
I tend to agree with @ajalab .